### PR TITLE
Debug CanvasItem redraw

### PR DESCRIPTION
**Pull request by reduz**
I wanted to add this tool for years and always forget. This command line option:

```
$ godot.exe -e --debug-canvas-item-redraw
```

Allows to see when a canvas item is redrawn. This helps find out if something in the UI is refreshing in a way it should not. Examples as such:

* Signals causing more of the UI to redraw.
* Container resizing causes more UI elements to redraw.
* Something using a timer is redrawing all time time, which can go unnoticed.

To my surprise, the editor UI is redrawing very efficiently. There is some weird stuff with the scene tabs, redrawing when the inspector changes but most things for the most part are fine.

How it looks:

debug-scene-redraw

This can be enabled also from the Editor UI to debug canvas item redraws in the running game:

image

### CODE REVIEW

**Comment by Calinou** [comment]
I suggest exposing this in the **Debug** menu at the top of the editor :slightly_smiling_face:

**Comment by reduz** [comment]
@Calinou oh, that's a very good point.

**Comment by Zireael07** [comment]
Would be awesome for editor plugins and tool scripts

**Comment by reduz** [comment]
@Calinou done

**Comment by clayjohn** [comment]
Honestly, I would name it differently. "Debug CanvasItem redraw" makes it sound like it is going to highlight each time the CanvasItem is drawn to the screen. When in fact, it is showing when the CanvasItem is recalculating it's properties on the CPU-side. Regardless of whether the CanvasItem recalculates it's properties, it is drawn to the screen.

File: main/main.cpp [hunk 1]
```diff
@@ -461,6 +462,7 @@ void Main::print_help(const char *p_binary) {
 	OS::get_singleton()->print("  --debug-navigation                Show navigation polygons when running the scene.\n");
 	OS::get_singleton()->print("  --debug-avoidance                 Show navigation avoidance debug visuals when running the scene.\n");
 	OS::get_singleton()->print("  --debug-stringnames               Print all StringName allocations to stdout when the engine quits.\n");
+	OS::get_singleton()->print("  --debug-canvas-item-redraw        Debug every time a canvas item redraws.\n");
```
Comment (Calinou) [inline main/main.cpp:465]:
```suggestion
	OS::get_singleton()->print("  --debug-canvas-item-redraw        Display a rectangle each time a canvas item redraws. Use this to troubleshoot low processor usage mode.\n");
```

File: editor/plugins/debugger_editor_plugin.cpp [hunk 2]
```diff
@@ -81,6 +81,10 @@ DebuggerEditorPlugin::DebuggerEditorPlugin(PopupMenu *p_debug_menu) {
 	debug_menu->set_item_tooltip(-1,
 			TTR("When this option is enabled, avoidance objects shapes, radius and velocities will be visible in the running project."));
 	debug_menu->add_separator();
+	debug_menu->add_check_shortcut(ED_SHORTCUT("editor/visible_canvas_redraw", TTR("Debug CanvasItem Redraws")), RUN_DEBUG_CANVAS_REDRAW);
+	debug_menu->set_item_tooltip(-1,
+			TTR("When this option is enabled, redraws of 2D objects will become visible (as a short flash) in the running project."));
```
Comment (Calinou) [inline editor/plugins/debugger_editor_plugin.cpp:86]:
```suggestion
			TTR("When this option is enabled, redraws of 2D objects will become visible (as a short flash) in the running project.\nThis is useful to troubleshoot low processor mode."));
```

**Comment by reduz** [comment]
@clayjohn No idea, should I call it Debug GUI/2D Node Redraw ? I have no idea what might be better.

**Comment by clayjohn** [comment]
> @clayjohn No idea, should I call it Debug GUI/2D Node Redraw ? I have no idea what might be better.

Perhaps. That is slightly better. 

I guess the real problem is that this type of update is called "redraw" internally. But "redraw" really just means "re-record" commands. Drawing happens every single frame regardless of whether the CanvasItem is "redrawn"

**Comment by akien-mga** [comment]
To me it's fairly established in Godot lingo that "redraw" indeed means "run the `update()` method to recompute what should be drawn", and that's the expensive part that needs to be minimized.

**Review by Calinou — APPROVED** [review]
Tested locally, it works as expected.

File: drivers/gles3/rasterizer_canvas_gles3.h [hunk 3]
```diff
@@ -367,6 +367,8 @@ class RasterizerCanvasGLES3 : public RendererCanvasRender {
 
 	void set_time(double p_time);
 
+	virtual void set_debug_redraw(bool p_enabled, double p_time, const Color &p_color) override {}
```
Comment (kleonc) [inline drivers/gles3/rasterizer_canvas_gles3.h:370]: Might be a good idea to mention somewhere (in the toggle's tooltip?:thinking:) that currently it's not implemented for GLES3.

Comment (clayjohn) [reply drivers/gles3/rasterizer_canvas_gles3.h:370]: Good thinking, maybe a warning would be sufficient for now

```suggestion
	virtual void set_debug_redraw(bool p_enabled, double p_time, const Color &p_color) override {
		if (p_enabled) {
			WARN_PRINT_ONCE("Debug CanvasItem Redraw is not available yet when using the GL Compatibility backend.);
		}
	}
```

Comment (kleonc) [reply drivers/gles3/rasterizer_canvas_gles3.h:370]: A warning makes sense, should be more informative/discoverable than a tooltip indeed. :+1:

**Comment by KoBeWi** [comment]
This does not work well with update spinner, it's constantly redrawing:

https://github.com/godotengine/godot/assets/2223172/b0134649-c4a4-4399-a51f-b1c3b083de04

Using `--debug-canvas-item-redraw` should probably force-disable it.

**Comment by akien-mga** [comment]
Rebased to solve merge conflict, and added the warning suggested by @clayjohn.

File: servers/rendering/renderer_rd/renderer_canvas_render_rd.cpp [hunk 4]
```diff
@@ -957,7 +957,48 @@ void RendererCanvasRenderRD::_render_item(RD::DrawListID p_draw_list, RID p_rend
...
 	}
+#ifdef DEBUG_ENABLED
+	if (debug_redraw && p_item->debug_redraw_time > 0.0) {
+		Color dc = debug_redraw_color;
+		dc.a *= p_item->debug_redraw_time / debug_redraw_time;
 
+		RID pipeline = pipeline_variants->variants[PIPELINE_LIGHT_MODE_DISABLED][PIPELINE_VARIANT_QUAD].get_render_pipeline(RD::INVALID_ID, p_framebuffer_format);
+		RD::get_singleton()->draw_list_bind_render_pipeline(p_draw_list, pipeline);
+
+		//bind textures
```
Comment (AThousandShips) [inline servers/rendering/renderer_rd/renderer_canvas_render_rd.cpp:968]:
```suggestion
		// Bind textures.
```

Comment (akien-mga) [reply servers/rendering/renderer_rd/renderer_canvas_render_rd.cpp:968]: I saw this, but decided not to change it. The whole file needs a cleanup pass:
```
servers/rendering/renderer_rd/renderer_canvas_render_rd.cpp:				//bind pipeline
servers/rendering/renderer_rd/renderer_canvas_render_rd.cpp:				//bind textures
servers/rendering/renderer_rd/renderer_canvas_render_rd.cpp:				//bind pipeline
servers/rendering/renderer_rd/renderer_canvas_render_rd.cpp:				//bind textures
servers/rendering/renderer_rd/renderer_canvas_render_rd.cpp:				//bind pipeline
servers/rendering/renderer_rd/renderer_canvas_render_rd.cpp:				//bind textures
servers/rendering/renderer_rd/renderer_canvas_render_rd.cpp:				//bind pipeline
servers/rendering/renderer_rd/renderer_canvas_render_rd.cpp:				//bind textures
```

**Review by clayjohn — APPROVED** [review]
Looks good to me. Testing locally it appears to work well.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**